### PR TITLE
web: add expand-all and collapse-all buttons

### PR DESCRIPTION
**Pull request by nicks**
Hello @landism, @lizzthabet,

Please review the following commits I made in branch nicks/ch13280:

7d6cdc6406cb6471a44ae9488dc462e431adda1a (2022-03-21 18:14:18 -0400)
web: add expand-all and collapse-all buttons

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
Screenshot from 2022-03-21 18-15-05

File: web/src/ResourceGroupsContext.tsx [hunk 1]
```diff
@@ -67,10 +71,25 @@ export function ResourceGroupsContextProvider(
...
+    // We can expand all groups by resetting the collapse state to empty."
+    function expandAll() {
+      setGroups({}) // Reset state.
+    }
+
+    // To collapse all groups, we need to know all their names.
+    function collapseAll(groupNames: string[]) {
+      let newState: GroupsState = {}
+      groupNames.forEach((group) => (newState[group] = { expanded: false }))
+      setGroups(newState)
```
Comment (lizzthabet) [inline web/src/ResourceGroupsContext.tsx:84]: this is a minor detail / edge case that i'm not sure is worth addressing, but i'll mention it anyways! when disabled resources are hidden, label groups with _only_ disabled resources will be reset to expanded when `collapseAll` is invoked.

**Review by lizzthabet — APPROVED** [review]
this looks great! are you planning to add the expand and collapse all buttons to detail view in a separate pr?

File: web/src/ResourceGroupsContext.tsx (hunk 1 above)
Comment (nicks) [reply web/src/ResourceGroupsContext.tsx:84]: oh boy it gets more complicated than that!

- if you click "collapse all", and then create new groups, do they default to expanded or collapsed?
- if you remove a collapsed group, click "expand all", and then bring the group back, does it default to expanded or collapsed?

i don't think there's an obvious right answer here...i like how the existing model encourages a default state of "expanded" when there are any open questions.

will leave some comments on this.

**Comment by nicks** [comment]
oh, i didn't realize that we were planning to add this to the sidebar too! ya, let's do that in a follow-up.

**Comment by nicks** [comment]
https://app.shortcut.com/windmill/story/13409/add-expand-collapse-buttons-to-sidebar?vc_group_by=day&ct_workflow=all&cf_workflow=500000005

also - i hope i didn't step on your toes! didn't realize there was a duplicate ticket :grimacing:

**Comment by milas** [comment]
Old duplicate ticket ([ch12977](https://app.shortcut.com/windmill/story/12977/nice-to-have-add-expand-and-collapse-groups-buttons-to-both-views)) is largely my fault!

I assigned it to myself with the intent of doing it a bit back but ended up never getting to it, so no worries, and glad this made it in regardless!

**Pull request merged**